lib/srv/db/mysql/gcp.go

```diff
@@ -124,52 +179,129 @@ func (e *Engine) getGCPUserAndPassword(ctx context.Context, sessionCtx *common.S
+func (a *gcpAuth) getGCPIAMAuthToken(ctx context.Context, dbUser string) (string, error) {
+	a.log.DebugContext(ctx, "Authenticating GCP MySQL with IAM auth.", "db_user", dbUser)
 
 	// Note that sessionCtx.DatabaseUser is the service account.
-	password, err := e.Auth.GetCloudSQLAuthToken(ctx, sessionCtx.DatabaseUser)
+	password, err := a.auth.GetCloudSQLAuthToken(ctx, dbUser)
 	return password, trace.Wrap(err)
 }
 
-func (e *Engine) getGCPOneTimePassword(ctx context.Context, sessionCtx *common.Session) (string, error) {
-	e.Log.DebugContext(ctx, "Authenticating GCP MySQL with password auth.", "session", sessionCtx)
+func (a *gcpAuth) getGCPOneTimePassword(ctx context.Context) (string, error) {
+	a.log.DebugContext(ctx, "Authenticating GCP MySQL with password auth")
 
 	// For Cloud SQL MySQL legacy auth, we use one-time passwords by resetting
 	// the database user password for each connection. Thus, acquire a lock to
 	// make sure all connection attempts to the same database and user are
 	// serialized.
 	retryCtx, cancel := context.WithTimeout(ctx, defaults.DatabaseConnectTimeout)
 	defer cancel()
-	lease, err := services.AcquireSemaphoreWithRetry(retryCtx, e.makeAcquireSemaphoreConfig(sessionCtx))
+	lease, err := services.AcquireSemaphoreWithRetry(retryCtx, a.makeAcquireSemaphoreConfig())
 	if err != nil {
 		return "", trace.Wrap(err)
 	}
 	// Only release the semaphore after the connection has been established
 	// below. If the semaphore fails to release for some reason, it will
 	// expire in a minute on its own.
 	defer func() {
-		err := e.AuthClient.CancelSemaphoreLease(ctx, *lease)
+		err := a.authClient.CancelSemaphoreLease(ctx, *lease)
 		if err != nil {
-			e.Log.ErrorContext(ctx, "Failed to cancel lease.", "lease", lease, "error", err)
+			a.log.ErrorContext(ctx, "Failed to cancel lease.", "lease", lease, "error", err)
 		}
 	}()
-	password, err := e.Auth.GetCloudSQLPassword(ctx, sessionCtx.Database, sessionCtx.DatabaseUser)
+	password, err := a.auth.GetCloudSQLPassword(ctx, a.database, a.databaseUser)
 	if err != nil {
 		return "", trace.Wrap(err)
 	}
 	return password, nil
 }
 
-const (
-	// gcpMySQLDBUserTypeBuiltIn indicates the database's built-in user type.
-	gcpMySQLDBUserTypeBuiltIn = "BUILT_IN"
-	// gcpMySQLDBUserTypeServiceAccount indicates a Cloud IAM service account.
-	gcpMySQLDBUserTypeServiceAccount = "CLOUD_IAM_SERVICE_ACCOUNT"
-	//  gcpMySQLDBUserTypeGroupServiceAccount indicates a Cloud IAM group service account.
-	gcpMySQLDBUserTypeGroupServiceAccount = "CLOUD_IAM_GROUP_SERVICE_ACCOUNT"
-	// gcpMySQLDBUserTypeUser indicates a Cloud IAM user.
-	gcpMySQLDBUserTypeUser = "CLOUD_IAM_USER"
-	// gcpMySQLDBUserTypeGroupUser indicates a Cloud IAM group login user.
-	gcpMySQLDBUserTypeGroupUser = "CLOUD_IAM_GROUP_USER"
-)
+// makeAcquireSemaphoreConfig builds parameters for acquiring a semaphore
+// for connecting to a MySQL Cloud SQL instance for this session.
```

Comment:
i've created a ticket to deprecate this:
https://github.com/gravitational/teleport/issues/61786